temporalcli/commands.workflow_reset.go

```diff
@@ -0,0 +1,205 @@
+
+	resetBaseRunID := c.RunId
+	eventID := int64(c.EventId)
+	if c.Type.Value != "" {
+		resetBaseRunID, eventID, err = c.getResetEventIDByType(cctx, cl)
+		if err != nil {
+			return fmt.Errorf("getting reset event ID by type failed: %w", err)
+		}
+	}
+	reapplyType := enums.RESET_REAPPLY_TYPE_SIGNAL
+	if c.ReapplyType.Value != "All" {
+		reapplyType, err = enums.ResetReapplyTypeFromString(c.ReapplyType.Value)
+		if err != nil {
+			return err
+		}
+	}
+
+	cctx.Printer.Printlnf("Resetting workflow %s to event ID %d", c.WorkflowId, eventID)
+
+	resp, err := cl.ResetWorkflowExecution(cctx, &workflowservice.ResetWorkflowExecutionRequest{
+		Namespace: c.Parent.Namespace,
+		WorkflowExecution: &common.WorkflowExecution{
+			WorkflowId: c.WorkflowId,
+			RunId:      resetBaseRunID,
+		},
+		Reason:                    fmt.Sprintf("%s: %s", username(), c.Reason),
+		WorkflowTaskFinishEventId: eventID,
+		ResetReapplyType:          reapplyType,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to reset workflow: %w", err)
+	}
+
+	if cctx.JSONOutput {
+		return cctx.Printer.PrintStructured(
+			resp,
+			printer.StructuredOptions{})
+	}
+	return nil
+}
+
+func (c *TemporalWorkflowResetCommand) getResetEventIDByType(ctx context.Context, cl client.Client) (string, int64, error) {
+	resetType, namespace, wid, rid := c.Type.Value, c.Parent.Namespace, c.WorkflowId, c.RunId
+	wfsvc := cl.WorkflowService()
+	switch resetType {
+	case "LastWorkflowTask":
+		return getLastWorkflowTaskEventID(ctx, namespace, wid, rid, wfsvc)
+	case "LastContinuedAsNew":
+		return getLastContinueAsNewID(ctx, namespace, wid, rid, wfsvc)
+	case "FirstWorkflowTask":
+		return getFirstWorkflowTaskEventID(ctx, namespace, wid, rid, wfsvc)
+	default:
+		return "", -1, fmt.Errorf("invalid reset type: %s", resetType)
+	}
+}
+
+// Returns event id of the last completed task or id of the next event after scheduled task.
+func getLastWorkflowTaskEventID(ctx context.Context, namespace, wid, rid string, wfsvc workflowservice.WorkflowServiceClient) (resetBaseRunID string, workflowTaskEventID int64, err error) {
+	resetBaseRunID = rid
+	req := workflowservice.GetWorkflowExecutionHistoryReverseRequest{
+		Namespace: namespace,
+		Execution: &common.WorkflowExecution{
+			WorkflowId: wid,
+			RunId:      rid,
+		},
+		MaximumPageSize: 250,
+		NextPageToken:   nil,
+	}
+
+	for more := true; more; more = len(req.NextPageToken) != 0 {
+		resp, err := wfsvc.GetWorkflowExecutionHistoryReverse(ctx, &req)
+		if err != nil {
+			return "", 0, fmt.Errorf("failed to get workflow execution history: %w", err)
+		}
+		for _, e := range resp.GetHistory().GetEvents() {
+			if e.GetEventType() == enums.EVENT_TYPE_WORKFLOW_TASK_COMPLETED {
+				workflowTaskEventID = e.GetEventId()
+				break
+			} else if e.GetEventType() == enums.EVENT_TYPE_WORKFLOW_TASK_SCHEDULED {
+				// if there is no task completed event, set it to first scheduled event + 1
+				workflowTaskEventID = e.GetEventId() + 1
+			}
+		}
```

Comment:
This `for` loop is premature optimization. You could define a function: 
```
func getFirstEventOfType(events, eventType) event{
   for _, e := range events {
       if e.GetEventType() == eventType {
          return e
      }
  }
}
```
and then write code as follow:
```
    events = resp.GetHistory().getEvents()
    event = getFirstEventOfType(events, enums.EVENT_TYPE_WORKFLOW_TASK_COMPLETED)
    if event == nil {
        event = getFirstEventOfType(events, enums.EVENT_TYPE_WORKFLOW_TASK_SCHEDULED)
    } 
```
and in the `getLastContinuedAs new:
```
   event = getFirstEventOfType(reverse(events), enums.EVENT_TYPE_WORKFLOW_TASK_COMPLETED)
```

Reply:
`reverse(events)` is not the same as iterating in reverse order as this is a paginated query.